mitmproxy/protocol/socks_client.py

```diff
@@ -0,0 +1,82 @@
+import socket
+
+from netlib import socks
+from netlib.exceptions import TcpException
+from netlib import tcp
+
+from .base import Layer
+from ..exceptions import ProtocolException
+
+class SocksClientLayer(Layer):
+
+    def __init__(self, ctx, server_tls):
+        Layer.__init__(self, ctx)
+        self.server_tls = server_tls
+        self.socks_addr = tcp.Address(("127.0.0.1", 1080))
+        self.socks_username = "test"
+        self.socks_password = "test"
+
+    def __call__(self):
+        layer = self.ctx.next_layer(self)
+        layer()
+
+    def _socks_username_password_auth(self, rfile, wfile):
+        if self.socks_username is None or self.socks_password is None:
+            raise socks.SocksError(0, "authentication failed")
+        cauth = socks.UsernamePasswordAuth(socks.USERNAME_PASSWORD_VERSION.DEFAULT, self.socks_username, self.socks_password)
+        cauth.to_file(wfile)
+        wfile.flush()
+        sauth = socks.UsernamePasswordAuthResponse.from_file(rfile)
+        if sauth.status != 0:
+            raise socks.SocksError(0, "authentication failed")
+
+    def _get_atype(self, address):
+        try:
+            if address.use_ipv6:
+                taddr = socket.inet_pton(socket.AF_INET6, address.host)
```

Comment:
Did you check if that works without false-positives? We use `ipaddress.IPv4Address(self.addr.host)` in `netlib.socks`, this could be a nasty sources of inconsistencies.